dali/pipeline/operators/paste/bbox_paste.cc

```diff
@@ -0,0 +1,98 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operators/paste/bbox_paste.h"
+
+namespace dali {
+
+DALI_SCHEMA(BBoxPaste)
+    .DocStr(
+        R"code(Transforms bounding boxes so that they are in the same place in the image after pasting it onto a larger canvas.
+Corner coordinates: (x', y') = (x/ratio + paste_x', y/ratio + paste_y')
+Box sizes:          (w', h') = (w/ratio, h/ratio)
+Where     paste_x' = paste_x * (ratio - 1)/ratio
+          paste_y' = paste_y * (ratio - 1)/ratio
+Paste coordinates are normalized so that (0,0) aligns the image to top-left of the canvas and (1,1) aligns it to bottom-right.
+)code")
+  .NumInput(1)
+  .NumOutput(1)
+  .AllowMultipleInputSets()
+  .AddArg("ratio",
+      R"code(Ratio of canvas size to input size, must be > 1.)code",
+      DALI_FLOAT, true)
+  .AddOptionalArg("ltrb",
+              R"code(True, for two-point (ltrb).
+False for for width-height representation. Default: False)code",
+              false, false)
+  .AddOptionalArg("paste_x",
+      R"code(Horizontal position of the paste in image coordinates (0.0 - 1.0))code",
+      0.5f, true)
+  .AddOptionalArg("paste_y",
+      R"code(Vertical position of the paste in image coordinates (0.0 - 1.0))code",
+      0.5f, true);
+
+template<>
+void BBoxPaste<CPUBackend>::RunImpl(BackendWorkspace *ws, const int idx) {
+  const auto &input = ws->Input<CPUBackend>(idx);
+  const auto input_data = input.data<float>();
+
+  DALI_ENFORCE(input.type().id() == DALI_FLOAT, "Bounding box in wrong format");
+  DALI_ENFORCE(input.size() % 4 == 0, "Bounding box must consist of exactly 4 float coordinates");
+
+  auto *output = ws->Output<CPUBackend>(idx);
+  output->set_type(TypeInfo::Create<float>());
+  output->ResizeLike(input);
+  auto *output_data = output->mutable_data<float>();
+
+  int i = ws->data_idx();
+  // pasting onto a larger canvas scales bounding boxes down by scale ratio
+  float ratio = spec_.GetArgument<float>("ratio", ws, i);
+  float px = spec_.GetArgument<float>("paste_x", ws, i);
+  float py = spec_.GetArgument<float>("paste_y", ws, i);
+  float scale = 1 / ratio;
+
+  // offsets are scaled so that (0,0) pastes the image aligned to the top-left
+  // corner and (1,1) aligns it to the (bottom, right) corner
+  float ofs_mul = (ratio - 1) / ratio;
+  float ofsx = px * ofs_mul;
+  float ofsy = py * ofs_mul;
+
+  for (int j = 0; j + 4 <= input.size(); j += 4) {
+
+    auto x0 = input_data[j];
+    auto y0 = input_data[j + 1];
+    auto x1w = input_data[j + 2];
+    auto y1h = input_data[j + 3];
+
+    x0 = x0 * scale + ofsx;
+    y0 = y0 * scale + ofsy;
+    if (useLTRB) {
+      x1w = x1w * scale + ofsx;
+      y1h = y1h * scale + ofsy;
+    } else {
+      x1w = x1w * scale;
+      y1h = y1h * scale;
+    }
+    // CAVEAT: Precision issues may drive bounding box outside (0-1).
```

Comment:
If this happens, RandomBBoxCrop will crash. Would be good to align on that and decide what is our assumption about bounding boxes. If we only support [0,1] (like atm) these values have to be clamped

Reply:
I need to dig into it a bit, but it seems it's not a matter of simple clamping (with LTRB it is, but not with width/height). My educated guess is that it'd require some nextafter calls to be sure that nothing breaks.